[WIP] Add mask2former

shivalikasingh95 opened the pull request:
Added Mask2former config and Pixel Decoder changes

shivalikasingh95 committed:
Mask2Former initial commit

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former

shivalikasingh95 committed:
Added Mask2FormerConfig and PixelDecoder changes

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former

alaradirik commented on docs/source/en/model_doc/maskformer.mdx:
```diff
@@ -10,62 +10,44 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-breaking changes to fix it in the future. If you see something strange, file a [Github Issue](https://github.com/huggingface/transformers/issues/new?assignees=&labels=&template=bug-report.md&title).
-
-</Tip>
+# mask2former
```
This is a good start but we create a new mdx file for each model, we aim to keep MaskFormer as it is and add Mask2Former it.

Could you copy paste this content to create `docs/source/en/model_doc/mask2former.mdx` and revert all changes to `docs/source/en/model_doc/maskformer.mdx`?

alaradirik commented on docs/source/en/model_doc/maskformer.mdx:
```diff
@@ -10,62 +10,44 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 ## Overview
 
-The MaskFormer model was proposed in [Per-Pixel Classification is Not All You Need for Semantic Segmentation](https://arxiv.org/abs/2107.06278) by Bowen Cheng, Alexander G. Schwing, Alexander Kirillov. MaskFormer addresses semantic segmentation with a mask classification paradigm instead of performing classic pixel-level classification.
+The mask2former model was proposed in [<INSERT PAPER NAME HERE>](<INSERT PAPER LINK HERE>) by <INSERT AUTHORS HERE>.
```
```suggestion
The Mask2Former model was proposed in [Masked-attention Mask Transformer for Universal Image Segmentation
](https://arxiv.org/abs/2112.01527) by Bowen Cheng, Ishan Misra, Alexander G. Schwing, Alexander Kirillov, Rohit Girdhar.
```

alaradirik commented on docs/source/en/model_doc/maskformer.mdx:
```diff
@@ -10,62 +10,44 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 The abstract from the paper is the following:
 
-*Modern approaches typically formulate semantic segmentation as a per-pixel classification task, while instance-level segmentation is handled with an alternative mask classification. Our key insight: mask classification is sufficiently general to solve both semantic- and instance-level segmentation tasks in a unified manner using the exact same model, loss, and training procedure. Following this  ...
+*<INSERT PAPER ABSTRACT HERE>*
```
```suggestion
*Image segmentation is about grouping pixels with different semantics, e.g., category or instance membership, where each choice of semantics defines a task. While only the semantics of each task differ, current research focuses on designing specialized architectures for each task. We present Masked-attention Mask Transformer (Mask2Former), a new architecture capable of addressing any image segmentation task (panoptic, instance or semantic). Its key components include masked attention, which extracts localized features by constraining cross-attention within predicted mask regions. In addition to reducing the research effort by at least three times, it outperforms the best specialized architectures by a significant margin on four popular datasets. Most notably, Mask2Former sets a new state-of-the-art for panoptic segmentation (57.8 PQ on COCO), instance segmentation (50.1 AP on COCO) and semantic segmentation (57.7 mIoU on ADE20K).*

```

alaradirik commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -87,6 +87,7 @@
         ("m2m_100", "M2M100Config"),
         ("marian", "MarianConfig"),
         ("maskformer", "MaskFormerConfig"),
+        ("maskformer", "Mask2FormerConfig"),
```
```suggestion
        ("mask2former", "Mask2FormerConfig"),
```

alaradirik commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -211,6 +212,7 @@
         ("lxmert", "LXMERT_PRETRAINED_CONFIG_ARCHIVE_MAP"),
         ("m2m_100", "M2M_100_PRETRAINED_CONFIG_ARCHIVE_MAP"),
         ("maskformer", "MASKFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP"),
+        ("maskformer", "MASK2FORMER_PRETRAINED_CONFIG_ARCHIVE_MAP"),
```
```suggestion
        ("mask2former", "MASK2FORMER_PRETRAINED_CONFIG_ARCHIVE_MAP"),
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_1). All of your documentation changes will be reflected on that endpoint.

alaradirik commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -340,6 +342,7 @@
         ("m2m_100", "M2M100"),
         ("marian", "Marian"),
         ("maskformer", "MaskFormer"),
+        ("maskformer", "mask2former"),
```
```suggestion
        ("mask2former", "Mask2Former"),
```

alaradirik commented on src/transformers/models/auto/feature_extraction_auto.py:
```diff
@@ -57,6 +57,7 @@
         ("layoutlmv3", "LayoutLMv3FeatureExtractor"),
         ("levit", "LevitFeatureExtractor"),
         ("maskformer", "MaskFormerFeatureExtractor"),
+        ("maskformer", "Mask2FormerFeatureExtractor"),
```
```suggestion
        ("mask2former", "Mask2FormerFeatureExtractor"),
```

alaradirik replied:
You can optionally skip creating a models/mask2former/feature_extraction_mask2former.py file if the preprocessing steps are the same as MaskFormers. In this case, this line would be:
`("mask2former", "MaskFormerFeatureExtractor")`

alaradirik commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -86,6 +86,7 @@
         ("m2m_100", "M2M100Model"),
         ("marian", "MarianModel"),
         ("maskformer", "MaskFormerModel"),
+        ("maskformer", "Mask2FormerModel"),
```
```suggestion
        ("mask2former", "Mask2FormerModel"),
```

alaradirik commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -381,6 +382,7 @@
     [
         # Model for Instance Segmentation mapping
         ("maskformer", "MaskFormerForInstanceSegmentation"),
+        ("maskformer", "Mask2FormerForInstanceSegmentation"),
```
```suggestion
        ("mask2former", "Mask2FormerForInstanceSegmentation"),
```

alaradirik commented on src/transformers/models/mask2former/__init__.py:
```diff
@@ -0,0 +1,74 @@
+            Mask2FormerModel,
+            Mask2FormerPreTrainedModel,
+        )
+
+
```
```suggestion
```

You can run the following lines to automatically check for bugs and apply style corrections:
`make fixup`

This is one of the commands run by CircleCI to tests whether everything is working as expected when you create a PR.

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -0,0 +1,229 @@
+    This is the configuration class to store the configuration of a [`Mask2FormerModel`]. It is used to instantiate a
+    Mask2Former model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the Mask2Former
+    [shivi/mask2former-segmentation-swin-large-ade](https://huggingface.co/shivi/mask2former-segmentation-swin-large-ade) architecture trained
```
Just a small note to transfer the repo to the Facebook organization on the hub before the PR is merged

alaradirik replied:
Another quick note, it's better to set the model that is most likely to be used (good performance, relatively lower memory requirements) as the default model. So the default model would  be either swin-base-coco or swin-base-ade.

shivalikasingh95 replied:
got it. Will keep this in mind.

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -0,0 +1,229 @@
+        importance_sample_ratio: Optional[float] = 0.75,
+        oversample_ratio: Optional[float] = 3.0,
+        common_stride: Optional[int] = 4,
+        encoder_in_features: Optional[List[str]] = ['res2', 'res3', 'res5'],
```
Excellent work over here! Just remember that all input arguments should be included in the class / function docstring (line 73)

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -0,0 +1,229 @@
+            decoder_config = AutoConfig.for_model(decoder_type, **decoder_config)
+
+        if pixel_decoder_config is None:
+            
```
You can run the `make fixup` command before committing to automatically perform style corrections (extra lines), detect bugs, etc. and to make sure that your code is in the format expected by the CI tests.

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -0,0 +1,229 @@
+        super().__init__(**kwargs)
+
+    @classmethod
+    def from_backbone_and_decoder_configs(
```
Since there is a new `pixel_decoder_config`, the function name and content should be changed to load all configs:

`from_backbone_decoder_pixel_decoder_configs()`

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -0,0 +1,229 @@
+        """Instantiate a [`Mask2FormerConfig`] (or a derived class) from a pre-trained backbone model configuration and DETR model
+        configuration.
+
+            Args:
```
Same here, we will need to pass in the `pixel_decoder_config`to initialize the model correctly

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -0,0 +1,229 @@
+            backbone_config=backbone_config.to_dict(),
+            decoder_config=decoder_config.to_dict(),
+            **kwargs,
+        )
```
```suggestion
        return cls(
            backbone_config=backbone_config.to_dict(),
            decoder_config=decoder_config.to_dict(),
            pixel_decoder_config=pixel_decoder_config.to_dict(),
            **kwargs,
        )
```

alaradirik commented on src/transformers/models/mask2former/convert_mask2former_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,737 @@
+# coding=utf-8
```
Just a note, this script will need to be modified after the model implementation is complete.

alaradirik commented on src/transformers/models/mask2former/feature_extraction_mask2former.py:
```diff
@@ -0,0 +1,649 @@
+
+logger = logging.get_logger(__name__)
+
+
```
Looks good!

Just to let you know, we have a copied from mechanism that allows us to keep track of all copies of a function or class across different models. This makes it much easier to make changes to all effected parts if we detect a bug or want to improve the code:

`# Copied from transformers.models.maskformer.feature_extraction_maskformer.MaskFormerFeatureExtractor`

shivalikasingh95 replied:
sure, have added the copied from statement for feature extractor now

alaradirik commented on tests/models/mask2former/test_modeling_mask2former.py:
```diff
@@ -0,0 +1,426 @@
+        # class_queries_logits
+        class_queries_logits = outputs.class_queries_logits
+        self.assertEqual(class_queries_logits.shape, (1, model.config.num_queries, model.config.num_labels + 1))
+        expected_slice = torch.tensor(
+            [
+                [1.6512e00, -5.2572e00, -3.3519e00],
+                [3.6169e-02, -5.9025e00, -2.9313e00],
+                [1.0766e-04, -7.7630e00, -5.1263e00],
+            ]
+        ).to(torch_device)
```
This will need to be updated after the model implementation is complete.

Other than this, I don't think there is any need to add new tests.

alaradirik commented on tests/models/mask2former/test_modeling_mask2former.py:
```diff
@@ -0,0 +1,426 @@
+        self.assertEqual(
+            masks_queries_logits.shape, (1, model.config.num_queries, inputs_shape[-2] // 4, inputs_shape[-1] // 4)
+        )
+        expected_slice = [
+            [-1.3737124, -1.7724937, -1.9364233],
+            [-1.5977281, -1.9867939, -2.1523695],
+            [-1.5795398, -1.9269832, -2.093942],
+        ]
+        expected_slice = torch.tensor(expected_slice).to(torch_device)
```
The expected result will need to be updated, please see my comment below.

alaradirik commented on tests/models/mask2former/test_modeling_mask2former.py:
```diff
@@ -0,0 +1,426 @@
+class Mask2FormerModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def model_checkpoints(self):
+        return "facebook/mask2former-swin-small-coco"
```
Just a note, this line will need to be updated to an equivalent Mask2Former checkpoint

alaradirik commented on tests/models/mask2former/test_modeling_mask2former.py:
```diff
@@ -0,0 +1,426 @@
+            )
+        )
+
+        expected_slice_hidden_state = torch.tensor(
+            [[-0.8422, -0.8434, -0.9718], [-1.0144, -0.5565, -0.4195], [-1.0038, -0.4484, -0.1961]]
+        ).to(torch_device)
+        self.assertTrue(
+            torch.allclose(
+                outputs.pixel_decoder_last_hidden_state[0, 0, :3, :3], expected_slice_hidden_state, atol=TOLERANCE
+            )
+        )
+
+        expected_slice_hidden_state = torch.tensor(
+            [[0.2852, -0.0159, 0.9735], [0.6254, 0.1858, 0.8529], [-0.0680, -0.4116, 1.8413]]
+        ).to(torch_device)
+        self.assertTrue(
+            torch.allclose(
+                outputs.transformer_decoder_last_hidden_state[0, :3, :3], expected_slice_hidden_state, atol=TOLERANCE
+            )
+        )
```
Same here, the expected values will need to be updated.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3208 @@
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length,
+            sequence_length)`. Attentions weights from Detr's decoder after the attention softmax, used to compute the
+            weighted average in the self-attention heads.
+    """
```
Looks good! You will just need to add `transformer_encoder_features` and `multi_scale_features` to the class argument docstrings over here.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3208 @@
+            fpn_features.append(output)
+        return fpn_features
+
+def multiscale_deformable_attention_core(value, value_spatial_shapes, sampling_locations, attention_weights):
```
A couple of pointers - we use type casting to make sure that the input parameters and output are of the expected type. It is also good practice to use more intuitive names for variable such that users can understand what each line does.

```suggestion
def multiscale_deformable_attention_core(
    value: torch.Tensor,
    value_spatial_shapes: torch.Tensor,
    sampling_locations: torch.Tensor,, 
    attention_weights: torch.FloatTensor,
) -> torch.FloatTensor:
```

shivalikasingh95 committed:
reverting maskformer documentation and adding new file for mask2former

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former

shivalikasingh95 committed:
Merge branch 'add-mask2former' of https://github.com/shivalikasingh95/transformers into add-mask2former

shivalikasingh95 committed:
Update src/transformers/models/auto/configuration_auto.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Update src/transformers/models/auto/configuration_auto.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Update src/transformers/models/auto/feature_extraction_auto.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Update src/transformers/models/auto/modeling_auto.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Update src/transformers/models/auto/modeling_auto.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Update src/transformers/models/mask2former/__init__.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
updated from_backbone_decoder_pixel_decoder_configs() classmethod

shivalikasingh95 committed:
Merge branch 'add-mask2former' of https://github.com/shivalikasingh95/transformers into add-mask2former

shivalikasingh95 committed:
added copied from statement for feature extractor

shivalikasingh95 committed:
Update src/transformers/models/mask2former/modeling_mask2former.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Update src/transformers/models/auto/configuration_auto.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 reviewed (commented):
Hi Alara...have fixed most of the things you mentioned except a few:
1. test file related comments - saving to fix in the end
2. checkpoint links on hub - will fix in end once all checkpoints are added to facebook org on HF hub
3. updating doc strings - will fix before next round of code review
4. variable names - will fix before next round of code review

alaradirik commented on docs/source/en/model_doc/mask2former.mdx:
```diff
@@ -0,0 +1,69 @@
+
+## Overview
+
+The mask2former model was proposed in [Masked-attention Mask Transformer for Universal Image Segmentation](https://arxiv.org/abs/2112.01527) by Bowen Cheng, Ishan Misra, Alexander G. Schwing, Alexander Kirillov, Rohit Girdhar.
```
```suggestion
The Mask2Former model was proposed in [Masked-attention Mask Transformer for Universal Image Segmentation](https://arxiv.org/abs/2112.01527) by Bowen Cheng, Ishan Misra, Alexander G. Schwing, Alexander Kirillov, Rohit Girdhar.
```

alaradirik commented on docs/source/en/model_doc/mask2former.mdx:
```diff
@@ -0,0 +1,69 @@
+## Overview
+
+The mask2former model was proposed in [Masked-attention Mask Transformer for Universal Image Segmentation](https://arxiv.org/abs/2112.01527) by Bowen Cheng, Ishan Misra, Alexander G. Schwing, Alexander Kirillov, Rohit Girdhar.
+Mask2Former is a single architecture for panoptic, instance and semantic segmentation.
```
Maybe we can expand the description a bit more later on?

```suggestion
Mask2Former is a unified framework for panoptic, instance and semantic segmentation.
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3213 @@
+    multi_scale_features: Optional[List[torch.FloatTensor]] = None
+
+
+# Copied from transformers.models.maskformer.modeling_maskformer.MaskFormerPixelDecoderOutput with MaskFormer->Mask2Former
```
The copied from statement would result in test failures over here. This is because this class is not completely the same as the MaskFormer version (transformer_encoder_features, multi_scale_features, doc string changes).

We only use the copied from statement if the class / function is completely identical.

shivalikasingh95 replied:
Right, noted! Thanks for mentioning. Will remove it.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3213 @@
+        # we use the last feature map
+        last_feature_projected = self.mask_projection(transformer_encoder_features[-1])
+
+        ##redefine??
```
I don't think `Mask2FormerPixelDecoderOutput `would need to be redefined but we'd need to output all hidden_states regardless of `output_hidden_states` is set to True. This is because the hidden states of the pixel decoder need to be fed into the transformer decoder.

shivalikasingh95 replied:
oh makes sense.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3213 @@
+        self.config = config
+        self.num_layers = len(config.depths)
+        self.num_features = int(config.embed_dim * 2 ** (self.num_layers - 1))
+
+        self.embeddings = Mask2FormerSwinEmbeddings(config)
+        self.encoder = Mask2FormerSwinEncoder(config, self.embeddings.patch_grid)
+
+        self.layernorm = nn.LayerNorm(self.num_features, eps=config.layer_norm_eps)
+        self.pooler = nn.AdaptiveAvgPool1d(1) if add_pooling_layer else None
+
+    def get_input_embeddings(self):
+        return self.embeddings.patch_embeddings
+
+    def _prune_heads(self, heads_to_prune):
+        """
+        Prunes heads of the model. heads_to_prune: dict of {layer_num: list of heads to prune in this layer} See base
+        class PreTrainedModel
+        """
+        for layer, heads in heads_to_prune.items():
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
+    def forward(
+        self,
+        pixel_values=None,
+        head_mask=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
```
The intermediate hidden states of the new pixel decoder will need to be fed to the transformer decoder (`Mask2FormerTransformerModule`).

shivalikasingh95 replied:
You mean it'll need to be fed into the Mask2FormerTransformerModule here, right?
Mask2FormerSwinModel is the backbone if I'm not mistaken

alaradirik replied:
Yes, exactly, sorry for the typo!

alaradirik reviewed (commented):
Hi again, the changes look good! I left a few comments to clarify what modules will need to be updated.

Apart from the other changes to the transformer decoder, the intermediate hidden states of the new pixel decoder will need to be fed into transformer decoder.

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former

shivalikasingh95 committed:
Update docs/source/en/model_doc/mask2former.mdx

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Update docs/source/en/model_doc/mask2former.mdx

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
1. Added Transformer Module changes for mask2former
2Added Deformable Detr Encoder classes from deformable_detr Implementation for pixel Decoder
Fixed Pixel Decoder Implementation

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former

shivalikasingh95 committed:
Added loss calculation changes and fixes for Mask2FormerInstanceSegmentation

shivalikasingh95 committed:
Merge branch 'main' of https://github.com/huggingface/transformers into add-mask2former

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -97,11 +97,11 @@ class Mask2FormerConfig(PretrainedConfig):
     backbones_supported = ["swin"]
     decoders_supported = ["detr"]
-    pixel_decoders_supported = ["detr"] #since deformable_detr not available in transformers yet
+    pixel_decoders_supported = ["deformable_detr"] #since deformable_detr not available in transformers yet
```
No need for this comment anymore:
```suggestion
    pixel_decoders_supported = ["deformable_detr"] 
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        
+        return Mask2FormerPixelLevelModuleOutput(
+            # the last feature is actually the output from the last layer
+            encoder_last_hidden_state=features[-1],
```
This is the output (encoded features) of the SWIN backbone. We don't need to return this anymore because it was used as the input to both the Pixel Decoder and Transformer Decoder in MaskFormer but is only fed to the Pixel Decoder in Mask2Former.

```suggestion
```

alaradirik replied:
Or maybe we can optionally return it?

shivalikasingh95 replied:
The intention behind returning it here was so that it can be passed to `Mask2FormerModelOutput` inside `Mask2FormerModel` since currently `Mask2FormerModelOutput` has outputs (hidden states) of backbone, pixel decoder and transformer modules.

If we don't return it here then should I remove `encoder_last_hidden_state` from `Mask2FormerModelOutput` as well?

also if we make it optional then can I do it based on value of `output_hidden_states` or should I do it on basis of a new parameter like `output_encoder_hidden_states`?

alaradirik replied:
Making it optional sounds good! The idea is to allow users to experiment with different things and make it easier to conduct research.

shivalikasingh95 replied:
okay makes sense. Will make it optional then.

shivalikasingh95 committed:
imports related merge fix

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+    Mask2Former's pixel level module output. It returns both the last and (optionally) the hidden states from the
+    `encoder` and `decoder`. By default, the `encoder` is a Mask2FormerSwin Transformer and the `decoder` is a 
+    MultiScaleDeformableAttention Transformer.
+
```
No need to return backbone / encoder output anymore so it should be optional, please see my comment in `Mask2FormerPixelLevelModule`.  Also, the pixel decoder needs to return all hidden states (multi-scale features) and it is no longer optional.
```suggestion
    Mask2Former's pixel level module output. It returns the output of the encoder (optional) and all hidden states (multi-scale features) from the `decoder`. By default, the `encoder` is a Mask2FormerSwin Transformer and the `decoder` is a 
    MultiScaleDeformableAttention Transformer.

```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+    MultiScaleDeformableAttention Transformer.
+
+    The `encoder_last_hidden_state` are referred on the paper as **images features**, while `decoder_last_hidden_state`
+    as **pixel embeddings**
```
Will need to be updated as well.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        encoder_hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each stage) of
+            shape `(batch_size, num_channels, height, width)`. Hidden-states (also called feature maps) of the model at
+            the output of each stage.
```
Would be redundant to return all hidden states of the backbone encoder (SWIN) since its output is no longer fed into the Transformer Decoder.

```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+
+    Args:
+        encoder_last_hidden_state (`torch.FloatTensor` of shape`(batch_size, num_channels, height, width)`):
+            Last hidden states (final feature map) of the last stage of the encoder.
```
```suggestion
        encoder_last_hidden_state (`torch.FloatTensor`, *optional*):
            Last hidden states (final feature map) of shape`(batch_size, num_channels, height, width)`, output of the last stage of the encoder.
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        decoder_hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings + one for the output of each stage) of
+            shape `(batch_size, num_channels, height, width)`. Hidden-states (also called feature maps) of the model at
+            the output of each stage.
```
These are no longer optional, please see my comments above.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+
+    encoder_last_hidden_state: Optional[torch.FloatTensor] = None
+    decoder_last_hidden_state: Optional[torch.FloatTensor] = None
+    encoder_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+    decoder_last_hidden_state: Optional[torch.FloatTensor] = None
+    encoder_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+    decoder_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+    multi_scale_features: Optional[List[torch.FloatTensor]] = None
```
`multi_scale_features` are hidden states extracted from the decoder, so it is the `decoder_hidden_states` and are no longer optional. 
```suggestion
    decoder_hidden_states: Tuple[torch.FloatTensor]
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+    last_hidden_state: torch.FloatTensor = None
+    transformer_encoder_features: torch.FloatTensor = None
+    multi_scale_features: List[torch.FloatTensor] = None
+    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
```
Sorry, I have overlooked this in my previous review. Please see my comments above, `multi_scale_features` is `hidden_states`.

```suggestion
    hidden_states: Tuple[torch.FloatTensor]
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        last_feature_projected = self.mask_projection(transformer_encoder_features[-1])
+
+
+        return Mask2FormerPixelDecoderOutput(
+            last_hidden_state=last_feature_projected,
+            transformer_encoder_features=transformer_encoder_features[0],
+            multi_scale_features=multi_scale_features,
+            hidden_states=tuple(transformer_encoder_features)
+        )
```
Please see my comments above, `multi_scale_features` is the `hidden_states`:
```suggestion
        return Mask2FormerPixelDecoderOutput(
            last_hidden_state=last_feature_projected,
            hidden_states=tuple(multi_scale_features)
        )
```

shivalikasingh95 replied:
sure fixing all changes related to this

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+
+class Mask2FormerPixelDecoder(nn.Module):
+    def __init__(self,
+        config: DeformableDetrConfig,
+        in_features: Dict = None,
+        feature_size: int = 256, 
+        mask_feature_size: int = 256,
+        feature_channels: List[int] = None,
+        feature_strides: List[int] = None,
+        common_stride: int = None):
```
Great work on the PR! But we aim to keep each module as self-contained as possible. For example, if we need to add a significant number of parameters to `DeformableDetrConfig` to define the Pixel Decoder, it'd be better to create a stand alone PixelDecoderConfig class in the configuration file and use it to initialize a more structured `MultiScaleDeformAttnEncoderModule`. My suggestions:

- Create a `PixelDecoderConfig` in configuration_mask2former.py instead of using `DeformableDetrConfig`.
- Rename `MultiScaleDeformAttnEncoderModule` to `Mask2FormerMSDAModel` to conform with the naming conventions.
- Move the content of this class and forward function to `Mask2FormerMSDAModel` such that you can initialize the specific pixel decoder and return the outputs (all hidden_states / multi_scale features)
self.fpn = MaskFormerFPNModel(*args, feature_size=feature_size, **kwargs)

shivalikasingh95 replied:
> For example, if we need to add a significant number of parameters to `DeformableDetrConfig` to define the Pixel Decoder, it'd be better to create a stand alone PixelDecoderConfig class in the configuration file and use it to initialize a more structured `MultiScaleDeformAttnEncoderModule`. 

Forgive me if this is a silly question but had a small doubt related to this suggested change. Since there are only 3 new parameters that I want to add to the existing `DeformableDetrConfig` to define the Pixel Decoder config...can I do something like shown below code snippet while initializing `Mask2FormerConfig` class?

```
self.pixel_decoder_config = pixel_decoder_config # where pixel_decoder_config is `DeformableDetrConfig` 
self.pixel_decoder_config.feature_strides = feature_strides
self.pixel_decoder_config.encoder_in_features = encoder_in_features
self.pixel_decoder_config.common_stride = common_stride
```
And then I can pass `Mask2FormerConfig.pixel_decoder_config` to the Pixel Decoder 

Or is creating a separate class for `PixelDecoderConfig` the recommended approach in such a scenario?

shivalikasingh95 replied:
Sorry again have a silly doubt here. Just wanted to double check to be sure.
The current hierarchy of classes is like this for the pixel decoder:
Mask2FormerPixelLevelModule -> Mask2FormerPixelDecoder() -> MultiScaleDeformAttnEncoderModule() -> DeformableDetrEncoder()

If I understand correctly, you're suggesting to combine `Mask2FormerPixelDecoder` and `MultiScaleDeformAttnEncoderModule` to form `Mask2FormerMSDAModel` so then the updated structure of classes for pixel decoder would be something like this:
Mask2FormerPixelLevelModule() -> Mask2FormerMSDAModel() -> DeformableDetrEncoder()

My only concern in the above scenario is that then the forward call of `Mask2FormerMSDAModel` might become a little lengthy. Would that be alright?

All questions are perfectly fine and to answer it,  no, I meant reorganizing the code such that:
Mask2FormerPixelLevelModule -> Mask2FormerPixelDecoder() -> Mask2FormerMSDAModel() (basically our version of original [`MSDeformAttnPixelDecoder` of Mask2Former](https://github.com/facebookresearch/Mask2Former/blob/main/mask2former/modeling/pixel_decoder/msdeformattn.py) -> submodules of `MSDeformAttnPixelDecoder`.

Does this clarify it?

alaradirik replied:
> Forgive me if this is a silly question but had a small doubt related to this suggested change. Since there are only 3 new parameters that I want to add to the existing `DeformableDetrConfig` to define the Pixel Decoder config...can I do something like shown below code snippet while initializing `Mask2FormerConfig` class?
> 
> ```
> self.pixel_decoder_config = pixel_decoder_config # where pixel_decoder_config is `DeformableDetrConfig` 
> self.pixel_decoder_config.feature_strides = feature_strides
> self.pixel_decoder_config.encoder_in_features = encoder_in_features
> self.pixel_decoder_config.common_stride = common_stride
> ```
> 
> And then I can pass `Mask2FormerConfig.pixel_decoder_config` to the Pixel Decoder

Yes, that is totally acceptable! We are just aiming for easy to understand and modularized code.

shivalikasingh95 replied:
> Does this clarify it?

Yes this clarifies it. Thanks!

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        return attn_output, attn_weights_reshaped
+
+#DetrDecoderLayer with reversed order of cross and self attention layers
+class DetrDecoderLayer(nn.Module):
```
I'm thinking it might be better to call this class `MaskedAttentionDecoder `or `TransformerDecoder` as it is not DETR anymore.

shivalikasingh95 replied:
Oh yes, sorry I forgot to rename the DetrDecoder classes. I think `MaskedAttentionDecoder` sounds good. Will fix this.

alaradirik reviewed (commented):
Great work on the PR! We are really excited to see this model getting added to transformers.

I think the main issues are due to naming of classes and structuring of the code. Mask2Former uses only a part DeformableDETR as its Pixel Decoder (MSDeformAttnPixelDecoder), which is defined [here](https://github.com/facebookresearch/Mask2Former/blob/main/mask2former/modeling/pixel_decoder/msdeformattn.py). So my suggestions are:
- Create a stand alone `PixelDecoderConfig` in configuration_mask2former.py to initialize the equivalent of MSDeformAttnPixelDecoder. Do a forward pass with both the original pixel decoder and your implementation to validate the input and output shapes and format.
- Mask2Former transformer decoder is fed the hidden states of the pixel decoder, `multi_scale_features` should be renamed to `hidden_states`.
- Rename `DetrDecoder` (transformer decoder) because it is not really DETR anymore, please see my code comments for suggestions. This new transformer decoder takes the hidden states (multi_scale_features) of the pixel decoder as input.

I believe it'll be much easier to finalize the model definition and restructuring and modularising the Pixel Decoder.

shivalikasingh95 commented:
> For example, if we need to add a significant number of parameters to `DeformableDetrConfig` to define the Pixel Decoder, it'd be better to create a stand alone PixelDecoderConfig class in the configuration file and use it to initialize a more structured `MultiScaleDeformAttnEncoderModule`. 

Had a small doubt related to this suggested change. Since there are only 3 new parameters that I want to add to the existing `DeformableDetrConfig` to define the Pixel Decoder config...can I do something like shown below code snippet while initializing `Mask2FormerConfig` class?

```
self.pixel_decoder_config = pixel_decoder_config # where pixel_decoder_config is `DeformableDetrConfig` 
self.pixel_decoder_config.feature_strides = feature_strides
self.pixel_decoder_config.encoder_in_features = encoder_in_features
self.pixel_decoder_config.common_stride = common_stride

And then I can pass `Mask2FormerConfig.pixel_decoder_config` to the Pixel Decoder 

Or is creating a separate class for PixelDecoderConfig the recommended approach in such a scenario?

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        labels (`torch.Tensor`):
+            A tensor with the same shape as inputs. Stores the binary classification labels for each element in inputs
+            (0 for the negative class and 1 for the positive class).
+        num_masks (`int`):
+            The number of masks present in the current batch, used for normalization.
```
Guessing this is just a left over argument.
```suggestion
```

shivalikasingh95 replied:
oh yes, it is.  thanks for catching.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+    
+        return outputs_mask, attention_mask
+
+
```
It'd be better to include a comment that refers to the original implementation, perhaps something like this:
```suggestion
# Adapted from https://github.com/facebookresearch/Mask2Former/blob/main/mask2former/modeling/matcher.py
```

shivalikasingh95 replied:
Sure

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+
+        Params:
+            cost_class (float, *optional*, defaults to 1.0):
+                This is the relative weight of the classification error in the matching cost.
```
Looks great, I just a have a few suggestions to make the docstrings a bit more concise.
```suggestion
                Relative weight of the classification error in the matching cost.
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        self.num_points = num_points
+
+    @torch.no_grad()
+    def forward(self, masks_queries_logits, class_queries_logits, mask_labels, class_labels) -> List[Tuple[Tensor]]:
```
```suggestion
    def forward(
        self, 
        masks_queries_logits: torch.Tensor, 
        class_queries_logits: torch.Tensor,
        mask_labels: torch.Tensor,
        class_labels: : torch.Tensor, 
) -> List[Tuple[Tensor]]:
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+
+    @torch.no_grad()
+    def forward(self, masks_queries_logits, class_queries_logits, mask_labels, class_labels) -> List[Tuple[Tensor]]:
+        """Performs the matching
```
```suggestion
        """
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            class_queries_logits (`torch.Tensor`):
+                A tensor` of dim `batch_size, num_queries, height, width` with the
+                  predicted masks.
+
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        preds_probs = class_queries_logits
+        
+        # iterate through batch size
+        for pred_probs, pred_mask, target_mask, labels in zip(preds_probs, preds_masks, mask_labels, class_labels):
```
These are logits, not final probabilities and masks so I think it'd be better to keep the naming as such. Also, this is entirely optional but it is preferable to keep things as concise as possible as long as it is still easy to follow:
```suggestion
        batch_size = mask_queries_logits.shape[0]
        for i in range(batch_size):
             pred_probs = class_queries_logits[i].softmax(-1)
             ...
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+
+            # Compute the classification cost. Contrary to the loss, we don't use the NLL,
+            # but approximate it in 1 - proba[target class].
+            # The 1 is a constant that doesn't change the matching, it can be ommitted.
```
FYI: You can fit this comment in a single line and then run make fixup for correct indentation /splitting of the sentence.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            # but approximate it in 1 - proba[target class].
+            # The 1 is a constant that doesn't change the matching, it can be ommitted.
+            cost_class = -pred_probs[:, labels]
+
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            target_mask = target_mask.to(pred_mask)
+            target_mask = target_mask[:, None]
+            pred_mask = pred_mask[:, None]
+
```
```suggestion

# Sample ground truth and predicted masks
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            target_mask = point_sample(
+                target_mask,
+                point_coordinates.repeat(target_mask.shape[0], 1, 1),
+                align_corners=Falsem
```
A small bug over here:
```suggestion
                align_corners=False,
```

shivalikasingh95 replied:
thanks for catching.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+                align_corners=Falsem
+            ).squeeze(1)
+
+            pred_mask = point_sample(
+                pred_mask,
+                point_coordinates.repeat(pred_mask.shape[0], 1, 1),
+                align_corners=False,
+            ).squeeze(1)
```
Another optional comment, you could make this a bit more concise like this:
```suggestion
            pred_coordinates = point_coordinates.repeat(pred_mask.shape[0], 1, 1)
            pred_mask = point_sample(pred_mask, pred_coordinates, align_corners=False).squeeze(1)
```

shivalikasingh95 replied:
oh yes, sure. Will do it. Thanks again.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        ]
+        return matched_indices
+
+    def __repr__(self):
+        head = "Matcher " + self.__class__.__name__
+        body = [
+            f"cost_class: {self.cost_class}",
+            f"cost_mask: {self.cost_mask}",
+            f"cost_dice: {self.cost_dice}",
+        ]
+        _repr_indent = 4
+        lines = [head] + [" " * _repr_indent + line for line in body]
+        return "\n".join(lines)
```
This was included in MaskFormer too but it is not needed as transformers have utility classes that can log losses and other metrics during training.
```suggestion
```

shivalikasingh95 replied:
oh okay thanks for letting me know. Will remove this then.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        return "\n".join(lines)
+
+
+class Mask2FormerLoss(nn.Module): #SetCriterion() in orig imp
```
```suggestion
# Adapted from https://github.com/facebookresearch/Mask2Former/blob/main/mask2former/modeling/criterion.py
class Mask2FormerLoss(nn.Module):
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+    def __init__(
+        self,
+        num_labels: int,
+        matcher: Mask2FormerHungarianMatcher,
```
No need to make this an argument, we can directly initialize it within __init__
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        super().__init__()
+        requires_backends(self, ["scipy"])
+        self.num_labels = num_labels
+        self.matcher = matcher
```
```suggestion
        self.matcher = Mask2FormerHungarianMatcher(
            num_points = num_points,
            ....,
        )
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        self.oversample_ratio = oversample_ratio
+        self.importance_sample_ratio = importance_sample_ratio
+
+    def _max_by_axis(self, the_list: List[List[int]]) -> List[int]:
```
A more descriptive argument name would be better, perhaps `sizes` instead of `the_list`?

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+                maxes[index] = max(maxes[index], item)
+        return maxes
+    
+    #nested_tensor_from_tensor_list() in orig imp
```
```suggestion
    # Adapted from nested_tensor_from_tensor_list() in original implementation
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        # compute finel size
+        batch_shape = [batch_size] + max_size
+        b, _, h, w = batch_shape
+        # get metadata
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            `Dict[str, Tensor]`: A dict of `torch.Tensor` containing the following key:
+            - **loss_cross_entropy** -- The loss computed using cross entropy on the predicted and ground truth labels.
+        """
+
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        pred_logits = class_queries_logits
+        batch_size, num_queries, _ = pred_logits.shape
+        criterion = nn.CrossEntropyLoss(weight=self.empty_weight)
+        idx = self._get_predictions_permutation_indices(indices)
+        # shape = (batch_size, num_queries)
```
Can fit to a single line
```suggestion
        idx = self._get_predictions_permutation_indices(indices # shape of (batch_size, num_queries)
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            )
+        return point_coordinates
+
+
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+
+    def __init__(self, in_features: int,  config: Mask2FormerConfig):
+        super().__init__()
+        #in_features -> in_channels
+        hidden_size = config.decoder_config.hidden_size #hidden_dim
+        print("hidden_size:",hidden_size)
+        print("in_features:",in_features)
+        should_project = in_features != hidden_size 
+        #self.pe_layer
+        self.position_embedder = Mask2FormerSinePositionEmbedding(num_pos_feats=hidden_size // 2, normalize=True)
+        #self.query_feat
+        self.learnable_queries = nn.Embedding(config.decoder_config.num_queries, hidden_size)
+        #self.query_embed
+        self.queries_embedder = nn.Embedding(config.decoder_config.num_queries, hidden_size)
+        
+        self.num_feature_levels = 3
+        self.level_embed = nn.Embedding(self.num_feature_levels, hidden_size)
+
+        self.input_projection = nn.ModuleList([
+            nn.Conv2d(in_features, hidden_size, kernel_size=1) 
```
Just leaving a note to clean up the comments later on.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            inputs_embeds=learnable_queries,
+            pixel_embeddings=pixel_embeddings,
+            feature_size_list=feature_size_list,
+            attention_mask=None,
```
No need to specify None values as the corresponding DetrDecoderOutput argument defaults to None:
```suggestion
```

shivalikasingh95 replied:
oh right, my bad.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+        print("queries_embeddings:",queries_embeddings.shape)
+        print("feature_size_list:",feature_size_list)
+
+        decoder_output: DetrDecoderOutput = self.decoder(
```
No need to typecast a second time:
```suggestion
        output = self.decoder(
```

shivalikasingh95 replied:
oh yes right, thanks for pointing it out.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            feature_size_list=feature_size_list,
+            attention_mask=None,
+            encoder_hidden_states=input_embeds,
+            encoder_attention_mask=None,
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            query_position_embeddings=queries_embeddings,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=None,
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -0,0 +1,3393 @@
+            return_dict=None,
+        )
+
+        return decoder_output
```
```suggestion
        return output
```

alaradirik reviewed (commented):
Hi Shiva, the loss functions look good to me and only need a few style changes and minor code clean ups! 

Just to clarify, is the new version of DetrDecoder equivalent to the [MultiScaleMaskedTransformerDecoder](https://github.com/facebookresearch/Mask2Former/blob/161be514814ea63440fae520d2ec72707b62ba54/mask2former/modeling/transformer_decoder/mask2former_transformer_decoder.py) of the original implementation?

If not, restructuring it would make it easier to move the mask prediction logic [over here](https://github.com/facebookresearch/Mask2Former/blob/161be514814ea63440fae520d2ec72707b62ba54/mask2former/maskformer_model.py) to the Mask2FormerForInstanceSegmentation class.

shivalikasingh95 committed:
bug fix

shivalikasingh95 committed:
updated doc string for pixel level module output

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 commented:
> Hi Shiva, the loss functions look good to me and only need a few style changes and minor code clean ups!
> 
> Just to clarify, is the new version of DetrDecoder equivalent to the [MultiScaleMaskedTransformerDecoder](https://github.com/facebookresearch/Mask2Former/blob/161be514814ea63440fae520d2ec72707b62ba54/mask2former/modeling/transformer_decoder/mask2former_transformer_decoder.py) of the original implementation?
> 
> If not, restructuring it would make it easier to move the mask prediction logic [over here](https://github.com/facebookresearch/Mask2Former/blob/161be514814ea63440fae520d2ec72707b62ba54/mask2former/maskformer_model.py) to the Mask2FormerForInstanceSegmentation class.

Yes @alaradirik  the new version of DetrDecoder is equivalent to the [MultiScaleMaskedTransformerDecoder](https://github.com/facebookresearch/Mask2Former/blob/161be514814ea63440fae520d2ec72707b62ba54/mask2former/modeling/transformer_decoder/mask2former_transformer_decoder.py) of the original implementation.
As far as I've understood the Mask2Former paper and code, the mask prediction logic needs to be part of the transformer module for the masked attention mechanism to work.
So I'm not sure if decoupling and moving the mask prediction logic to Mask2FormerForInstanceSegmentation class would be the right approach.
If you have any suggestions for this then please let me know.

shivalikasingh95 committed:
1. Restructured pixel decoder in modeling file
2. Added checkpoint conversion script for mask2former
3. Updated feature extractor for instance segmentation post processing
4. Doc string updates
5. config file fixes

shivalikasingh95 commented:
Hi @alaradirik pinging here too for review since you mentioned to ping here in last mail.

The current PR is updated with following changes:
1. Restructured pixel decoder in modeling file as suggested by you
2. Added checkpoint conversion script for mask2former
3. Updated feature extractor for instance segmentation post processing
4. Doc string updates
5. mask2former config file fixes as discussed during previous PR review.

Have shared a more detailed summary of the changes via email. Looking forward to your review.

alaradirik commented on src/transformers/models/mask2former/convert_mask2former_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -177,7 +184,7 @@ def __call__(self, original_config: object) -> Mask2FormerFeatureExtractor:
             max_size=model_input.MAX_SIZE_TEST,
             num_labels=model.SEM_SEG_HEAD.NUM_CLASSES,
-            ignore_index=dataset_catalog.ignore_label,
+            # ignore_index=dataset_catalog.ignore_label, #check
```
This is probably left-over code but why is `ignore_index `commented out?

shivalikasingh95 replied:
I was getting an error actually that's why I had commented this.  

AttributeError: Attribute 'ignore_label' does not exist in the metadata of dataset 'coco_2017_val'. Available keys are dict_keys(['name', 'json_file', 'image_root', 'evaluator_type', 'thing_dataset_id_to_contiguous_id', 'thing_classes', 'thing_colors'])

Needed to debug this -> that's why added a small `check` comment

alaradirik commented on src/transformers/models/mask2former/convert_mask2former_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -676,13 +853,13 @@ def get_name(checkpoint_file: Path):
-    from Mask2Former.mask_former import add_mask_former_config
-    from Mask2Former.mask_former.mask_former_model import Mask2Former as OriginalMask2Former
+    # import original Mask2Former config and model from original source code repo
+    from Mask2Former.mask2former.config import add_maskformer2_config
```
Maybe I'm missing it but it seems the `add_maskformer2_config` function is not defined anywhere.

This function should basically map the configuration variable names and values of the original model to their HF implementation version.

alaradirik committed:
update post process methods

alaradirik committed:
Merge branch 'huggingface:main' into add-mask2former

alaradirik committed:
add num_classes to configuration

shivalikasingh95 committed:
sync fork

shivalikasingh95 committed:
1. mask2former config - fixed default values
2. fixed mismatch in no. of params for decoder
3. fixed minor bug in FPN module

shivalikasingh95 committed:
bug fix

alaradirik committed:
fix ignore_index for conversion

alaradirik committed:
fix minor issues

alaradirik committed:
fix minor issues

alaradirik committed:
fix minor issues

alaradirik committed:
fix backbone issues

alaradirik committed:
reorganize Mask2FormerMSDAModel

alaradirik committed:
minor fixes

shivalikasingh95 committed:
1. fixed pixel decoder missing keys issue in model conversion file
2. removed extra layernorm in backbone
3. minor bug fix in pixel decoder MSDAModel

shivalikasingh95 committed:
Merge remote-tracking branch 'upstream/main' into add-mask2former

shivalikasingh95 committed:
pixel decoder module bug fixes

alaradirik committed:
add image processor

shivalikasingh95 committed:
masked attention decoder fix

shivalikasingh95 committed:
Merge branch 'add-mask2former' of https://github.com/shivalikasingh95/transformers into add-mask2former

shivalikasingh95 committed:
added pixeldecoder for gradient checkpointing

alaradirik committed:
fix preprocessing, postprocessing for conversion

alaradirik committed:
fix preprocessing, postprocessing for conversion

alaradirik committed:
Merge branch 'huggingface:main' into add-mask2former

alaradirik committed:
switch to using AutoBackbone

shivalikasingh95 committed:
Autobackbone change related bug fixes

NielsRogge committed:
Add Swin backbone (#20769)

* Add Swin backbone

* Remove line

* Add code example

Co-authored-by: Niels Rogge <nielsrogge@Nielss-MacBook-Pro.local>

shivalikasingh95 committed:
fix class_queries and mask_queries logits

Narsil committed:
Even more validation. (#20762)

* Even more validation.

* Fixing order.

ydshieh committed:
Install vision for TF pipeline tests (#20771)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

alaradirik committed:
restructure pixel decoder, transformer decoder

alaradirik committed:
rename model head to universal segmentation

alaradirik committed:
update tests and conversion script

larsmennen committed:
Patch for FlanT5-XXL 8bit support (#20760)

* Workaround for #20287: FlanT5-XXL 8bit support

* Make fix-copies

* revert unrelated change

* Dont apply to longt5 and switch transformers

younesbelkada committed:
[Pipeline] fix failing bloom `pipeline` test (#20778)

fix failing `pipeline` test

alaradirik committed:
fix bugs

Narsil committed:
Fixing object detection with `layoutlm` (#20776)

* Fixing object detection with layoutlm.

* Fixup.

ydshieh committed:
Install video dependency for pipeline CI (#20777)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

gante committed:
Generate: use `GenerationConfig` as the basis for `.generate()` parametrization (#20388)

* generate from config mvp

* fix failing tests

* max_time test

* Load default gen config at model load time; Update docs

* further documentation; add tests

* adapt rag to the new structure

* handle models not instantiated with from_pretained (like in tests)

* better default generation config

* add can_generate fn

* handle legacy use case of ad hoc model config changes

* initialize gen config from config in individual methods, if gen config is none

* fix _get_decoder_start_token_id when called outside GenerationMixin

* correct model config load order (set attr > model config > decoder config)

* update rag to match latest changes

* Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* load gen config from model config in model.from_pretrained

* fix can_generate fn

* handle generate calls without a previous from_pretrained (e.g. tests)

* add legacy behavior (and a warning)

* lower logger severity

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

amyeroberts committed:
Move convert_to_rgb to image_transforms module (#20784)

* Move convert_to_rgb to image_transforms module

* Fix tests

ydshieh committed:
Recompile `apex` in `DeepSpeed` CI image (#20788)

Recompile apex in DeepSpeed CI image

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

alaradirik committed:
fix bugs

alaradirik committed:
fix bugs

alaradirik committed:
Merge branch 'huggingface:main' into add-mask2former
